planner/core/find_best_task.go

```diff
@@ -418,6 +436,192 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
+		Ranges:           path.ranges,
+		dataSourceSchema: ds.schema,
+		isPartition:      ds.isPartition,
+		physicalTableID:  ds.physicalTableID,
+	}.Init(ds.ctx)
+	statsTbl := ds.statisticTable
+	if statsTbl.Indices[idx.ID] != nil {
+		is.Hist = &statsTbl.Indices[idx.ID].Histogram
+	}
+	is.initSchema(ds.id, idx, true)
+	rowCount := path.countAfterAccess
+	if prop.ExpectedCnt < ds.stats.RowCount {
+		selectivity := ds.stats.RowCount / path.countAfterAccess
+		rowCount = math.Min(prop.ExpectedCnt/selectivity, rowCount)
+	}
+	partialCost += rowCount * scanFactor
+	is.stats = property.NewSimpleStats(rowCount)
+	is.stats.StatsVersion = ds.statisticTable.Version
+	if ds.statisticTable.Pseudo {
+		is.stats.StatsVersion = statistics.PseudoVersion
+	}
+	indexConds := path.indexFilters
+	if indexConds != nil {
+		var selectivity float64
+		partialCost += rowCount * cpuFactor
+		if path.countAfterAccess > 0 {
+			selectivity = path.countAfterIndex / path.countAfterAccess
+		}
+		count := is.stats.RowCount * selectivity
+		stats := &property.StatsInfo{RowCount: count}
+		indexSel := PhysicalSelection{Conditions: indexConds}.Init(is.ctx, stats)
+		indexSel.SetChildren(is)
+		partialCost += count * netWorkFactor
+		return indexSel, partialCost, count, false
+	}
+	return is, partialCost, rowCount, false
+}
+
+func (ds *DataSource) convertToPartialTableScan(prop *property.PhysicalProperty, path *accessPath) (PhysicalPlan, float64, float64, bool) {
+	var partialCost float64
+	ts := PhysicalTableScan{
+		Table:           ds.tableInfo,
+		Columns:         ds.Columns,
+		TableAsName:     ds.TableAsName,
+		DBName:          ds.DBName,
+		isPartition:     ds.isPartition,
+		physicalTableID: ds.physicalTableID,
+	}.Init(ds.ctx)
+	ts.SetSchema(ds.schema)
+	if ts.Table.PKIsHandle {
+		if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+			if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+				ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+			}
+		}
+	}
+	ts.Ranges = path.ranges
+	ts.AccessCondition, ts.filterCondition = path.accessConds, path.tableFilters
+	rowCount := path.countAfterAccess
+	// Adjust number of rows we actually need to scan if prop.ExpectedCnt is smaller than the count we calculated.
+	if prop.ExpectedCnt < ds.stats.RowCount {
+		count, ok, corr := ds.crossEstimateRowCount(path, prop.ExpectedCnt, false)
+		if ok {
+			rowCount = count
+		} else if corr < 1 {
+			correlationFactor := math.Pow(1-corr, float64(ds.ctx.GetSessionVars().CorrelationExpFactor))
+			selectivity := ds.stats.RowCount / rowCount
+			rowCount = math.Min(prop.ExpectedCnt/selectivity/correlationFactor, rowCount)
+		}
+	}
+	ts.stats = property.NewSimpleStats(rowCount)
+	ts.stats.StatsVersion = ds.statisticTable.Version
+	if ds.statisticTable.Pseudo {
+		ts.stats.StatsVersion = statistics.PseudoVersion
+	}
+	partialCost += rowCount * scanFactor
+	partialCost += rowCount * netWorkFactor
+	return ts, partialCost, rowCount, false
+}
```

Comment:
Though currently `path.tableFilter` would be empty now because of the check in https://github.com/pingcap/tidb/blob/45790b5799d535dab0df7b7c6959d947be5f5fe1/planner/core/stats.go#L236, but actually we can relax the condition there to let table path have table filters, so we should consider the `Selection` above this partial table scan here.